end2end-test-examples/gcs/src/main/java/io/grpc/gcs/ZeroCopyMessageMarshaller.java

```diff
@@ -0,0 +1,112 @@
+package io.grpc.gcs;
+
+import com.google.protobuf.ByteString;
+import com.google.protobuf.CodedInputStream;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.MessageLite;
+import com.google.protobuf.Parser;
+import com.google.protobuf.UnsafeByteOperations;
+import io.grpc.Detachable;
+import io.grpc.HasByteBuffer;
+import io.grpc.KnownLength;
+import io.grpc.MethodDescriptor.PrototypeMarshaller;
+import io.grpc.protobuf.lite.ProtoLiteUtils;
+import io.grpc.Status;
+import io.grpc.MethodDescriptor;
+import java.io.InputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+
+/** Custom gRPC marshaller to use zero memory copy feature of gRPC when deserializing messages */
+class ZeroCopyMessageMarshaller<T extends MessageLite> implements PrototypeMarshaller<T> {
+  private Map<T, InputStream> unclosedStreams = Collections.synchronizedMap(new IdentityHashMap<>());
+  private final Parser<T> parser;
+  private final PrototypeMarshaller<T> baseMarshaller;
+
+  ZeroCopyMessageMarshaller(T defaultInstance) {
+    parser = (Parser<T>) defaultInstance.getParserForType();
+    baseMarshaller = (PrototypeMarshaller<T>) ProtoLiteUtils.marshaller(defaultInstance);
+  }
+
+  @Override
+  public Class<T> getMessageClass() {
+    return baseMarshaller.getMessageClass();
+  }
+
+  @Override
+  public T getMessagePrototype() {
+    return baseMarshaller.getMessagePrototype();
+  }
+
+  @Override
+  public InputStream stream(T value) {
+    return baseMarshaller.stream(value);
+  }
+
+  @Override
+  public T parse(InputStream stream) {
+    CodedInputStream cis = null;
+    try {
+      if (stream instanceof KnownLength) {
+        int size = stream.available();
+        if (stream instanceof Detachable
+            && stream instanceof HasByteBuffer
+            && ((HasByteBuffer) stream).byteBufferSupported()) {
+          // Stream is now detached here and should be closed later.
+          stream = ((Detachable) stream).detach();
+          // This mark call is to keep buffer while travelsing buffer using skip.
+          // No need to call counterpart reset because the stream isn't used but ByteBuffer inside.
```

Comment:
Ouch. I had probably not explained this clearly to you 😢 . It would still be better to call `stream.reset()` at the end of consumption.

It works without reset() because protobuf [slices the ByteBuffers](https://github.com/protocolbuffers/protobuf/blob/f71a662db4cce92f2fc53866e0fc27ae25177e9c/java/core/src/main/java/com/google/protobuf/NioByteString.java#L56) when doing the unsafe [wrap](https://github.com/protocolbuffers/protobuf/blob/f71a662db4cce92f2fc53866e0fc27ae25177e9c/java/core/src/main/java/com/google/protobuf/ByteString.java#L381). Therefore, even though bytes are skipped in ByteBuffers, they are still there and the sliced ByteBuffer [repositions the read pointer](https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#slice()):

> The new buffer's position will be zero, ...

That's why protobuf can still read bytes from ByteBuffers that have been skipped off. But really, we are kind of relying on the implementation detail of protobuf. So I'd say we would better do `stream.reset()` after the while loop. That can never go wrong.

Reply:
Oh yeah, it'd be better not to depend on the implementation detail. I updated the PR to call reset(). PTAL.